[#65401] Project phase field shows in pdf report even when it doesn't show on wp

as-op opened the pull request:
# Ticket

https://community.openproject.org/work_packages/65401

# What are you trying to achieve?

For exporting a PDF Attribute table and Macro workPackageValue:projectPhase need to take into account:
* permission for `:view_project_phase`
* min. one active project phase (to follow the work package form)

# What approach did you choose and why?

* To export a PDF attribute table, we need to ensure the user has permission and at least one active project phase is available.  This was achieved by checking permissions and the existence of at least one active project phases. 
* Add test cases (also for other workPackageValue, workPackageLabel, projectLabel, projectValues)

# Merge checklist

- [x] Added/updated tests
- [x] Tested major browsers (Chrome, Firefox, Edge, ...)

as-op committed:
[#65401] Project phase field shows in pdf report even when it doesn't show on wp

https://community.openproject.org/work_packages/65401

as-op committed:
pdf export: hide project phase attribute if there is no project phase active in the project

as-op committed:
do not throw errors in export for I18n::MissingTranslationData

as-op committed:
add specs for workPackageValue, workPackageLabel, projectLabel, projectValue macros

as-op committed:
add specs for workPackageValue:project_phase

as-op committed:
obey rubocop

as-op committed:
adjust spec setup

as-op committed:
obey rubocop

toy commented on app/models/work_package/pdf_export/export/wp/attributes.rb:
```diff
@@ -156,6 +160,10 @@ def group_attributes_group_parts(group, work_package)
 
+  def skip_attribute?(form_key, work_package)
+    !CustomField.custom_field_attribute?(form_key) && !user_allowed_view_attribute?(work_package, form_key)
+  end
```
```suggestion
  def show_attribute?(form_key, work_package)
    CustomField.custom_field_attribute?(form_key) || user_allowed_view_attribute?(work_package, form_key)
  end
```
or same as in other class `can_view_attribute?`

toy commented on app/models/work_package/exports/attributes.rb:
```diff
@@ -0,0 +1,46 @@
+        work_package.project.phases.active.any?
+    end
+
+    def user_allowed_view_attribute?(obj, attribute_name)
```
maybe better to call it `allowed_to_view_attribute?`

as-op replied:
done https://github.com/opf/openproject/pull/19449/commits/893165c135ff7cd22062a623511efc2c23c7e3dc

toy commented on app/models/work_package/exports/attributes.rb:
```diff
@@ -0,0 +1,46 @@
+
+module WorkPackage::Exports
+  module Attributes
+    def user_allowed_view_wp_project_phase?(work_package)
```
maybe better to call it `allowed_to_view_project_phases?` and accept `project` instead of `work_package`?

as-op replied:
done https://github.com/opf/openproject/pull/19449/commits/0fe4d90c7c85144c66f16946edc003c60b5e1315

toy commented on app/models/work_package/exports/macros/attributes.rb:
```diff
@@ -104,6 +105,10 @@ def self.resolve_label(model, attribute)
+      rescue I18n::MissingTranslationData
+        # If the translation is missing, we return the attribute name as a fallback
+        # # This can happen if the attribute is a custom field, the attribute is not translated or does not exist
+        attribute.to_s
```
Is this needed outside of tests? `raise_on_missing_translations` is true only for development and test

as-op replied:
Thx, I didn't know. I've removed the rescue code and disabled raise_on_missing_translations for the affected tests https://github.com/opf/openproject/pull/19449/commits/c1c7dc127f8e6ca4b26dd43ddbf8c113f6d28b55

toy commented on app/models/work_package/exports/macros/attributes.rb:
```diff
@@ -145,18 +150,30 @@ def self.resolve_value_work_package(work_package, attribute)
+
+      def self.convert_to_attribute_name(custom_field, attribute, obj)
+        if custom_field.nil?
+          ::API::Utilities::PropertyNameConverter.to_ar_name(attribute.to_sym, context: obj)
```
Same call in `resolve_label`, so maybe better to extract, but not sure

as-op replied:
done https://github.com/opf/openproject/pull/19449/commits/9831be871999e24ef6fc88e5a5234c311f8d428d

toy commented on app/models/work_package/exports/macros/attributes.rb:
```diff
@@ -145,18 +150,30 @@ def self.resolve_value_work_package(work_package, attribute)
+      end
+
+      def self.can_view_attribute?(custom_field, obj, attribute_name)
+        !(custom_field.nil? && !user_allowed_view_attribute?(obj, attribute_name))
```
```suggestion
        custom_field || user_allowed_view_attribute?(obj, attribute_name)
```

as-op replied:
done https://github.com/opf/openproject/pull/19449/commits/0b1b5f3168ab275b9fd48d921b76703f70e961be

toy commented on app/models/work_package/exports/macros/attributes.rb:
```diff
@@ -145,18 +150,30 @@ def self.resolve_value_work_package(work_package, attribute)
 
-        return msg_macro_error_rich_text if cf&.formattable?
+        attribute_name = convert_to_attribute_name(custom_field, attribute, obj)
+        return "" unless can_view_attribute?(custom_field, obj, attribute_name)
```
Does it need to be a string? Can it be `nil`:
```suggestion
        return unless can_view_attribute?(custom_field, obj, attribute_name)
```

as-op replied:
No, even more, empty string would be a possible bug in markdown, so I return a space character instead now.
https://github.com/opf/openproject/pull/19449/commits/38e924b0f3780d4637744ea8cd6d89a6644497f0

toy commented on spec/models/work_packages/pdf_export/work_package_to_pdf_spec.rb:
```diff
@@ -71,6 +71,10 @@
            # cf_disabled_in_project.id not included == disabled
            work_package_custom_field_ids: [cf_long_text.id, cf_empty_long_text.id, cf_global_bool.id, cf_link.id])
   end
+  let(:phase_definition) { create(:project_phase_definition, name: "Test Phase") }
+  let!(:project_phase) do
+    create(:project_phase, project: project, definition: phase_definition, active: true)
+  end
```
No need to explicitly create phase definition, it will be built/created automatically
```suggestion
  let!(:project_phase) do
    create(:project_phase, project: project, active: true)
  end
```

as-op replied:
removed https://github.com/opf/openproject/pull/19449/commits/2d414f6d387e08db966f00d54a29c946ee205ae2

toy commented on spec/models/work_package/pdf_export/common/macro_spec.rb:
```diff
@@ -31,14 +31,60 @@
+      work_package_custom_fields: [custom_field],
+      project_custom_fields: [project_custom_field],
+      custom_field_values: { project_custom_field.id => "Project custom value 1" },
+      &:save!
```
I think it will behave the same without explicit `save!`

as-op replied:
done
https://github.com/opf/openproject/pull/19449/commits/2d414f6d387e08db966f00d54a29c946ee205ae2

toy commented on spec/models/work_package/pdf_export/common/macro_spec.rb:
```diff
@@ -31,14 +31,60 @@
+      type: type_task,
+      status: create(:status, name: "In Progress"), project: project,
+      custom_field_values: { custom_field.id => "Custom value 1" },
+      &:save!
```
Same point about not needing `save!`

as-op replied:
done https://github.com/opf/openproject/pull/19449/commits/2d414f6d387e08db966f00d54a29c946ee205ae2

toy commented on spec/models/work_package/pdf_export/common/macro_spec.rb:
```diff
@@ -47,48 +93,399 @@
+
+      it "ignores the tag" do
+        expect(formatted).to eq("<mention class=\"mention\" data-id=\"185\" " +
+                                "data-type=\"work_package\" data-text=\"#185\">\\#185</mention>")
```
Cleaner with single quotes and no escaping in this and other places
```suggestion
        expect(formatted).to eq('<mention class="mention" data-id="185" ' +
                                'data-type="work_package" data-text="#185">#185</mention>')
```

as-op replied:
changed everywhere except where there is now a template string
https://github.com/opf/openproject/pull/19449/commits/81efe18f33001302d82d8b6c55fa808ecda766f9

toy commented on spec/models/work_package/pdf_export/common/macro_spec.rb:
```diff
@@ -47,48 +93,399 @@
+    end
+
+    describe "with markdown formating bold" do
+      let(:markdown) { "\n**#185**\n" }
```
Unclear, are new lines required for it to work?

as-op replied:
removed https://github.com/opf/openproject/pull/19449/commits/81efe18f33001302d82d8b6c55fa808ecda766f9

toy commented on spec/models/work_package/pdf_export/common/macro_spec.rb:
```diff
@@ -47,48 +93,399 @@
+    end
+
+    describe "with specific work package ID and attribute" do
+      let(:markdown) { "workPackageValue:185:subject" }
```
I think it is better to create a separate work package for this and other places checking value of specific work package

as-op replied:
done https://github.com/opf/openproject/pull/19449/commits/78d97f161d81d741e196c242b37fad1b880aac15

toy commented on spec/models/work_package/pdf_export/common/macro_spec.rb:
```diff
@@ -47,48 +93,399 @@
+      let(:markdown) { "workPackageValue:nonexistent_attribute" }
+
+      it "outputs an empty value" do
+        expect(formatted).to eq(" ")
```
I'd expect `""`

as-op replied:
I'm afraid not, as that could change the meaning of the Markdown.  For example, bold formatting might be interpreted as a horizontal line.

e.g. `**workPackageValue:nonexistent_attribute** `would be interpreted as `****` (horizontal line `***`)

toy commented on spec/models/work_package/pdf_export/common/macro_spec.rb:
```diff
@@ -31,14 +31,60 @@
+  shared_let(:work_package) do
+    create(
+      :work_package,
+      id: 185,
```
Better to leave it out and use interpolation in all usages with `work_package.id`

as-op replied:
done https://github.com/opf/openproject/pull/19449/commits/6e2d0e6d5e0f06128fb5c3591af671ec6a4f2b96

toy commented on spec/models/work_package/pdf_export/common/macro_spec.rb:
```diff
@@ -47,48 +93,399 @@
+    end
+
+    describe "with specific project ID and attribute" do
+      let(:markdown) { "projectValue:#{project.id}:name" }
```
In this and other places checking value of specific project better to use a different project to test that macro doesn't return value for this project

as-op replied:
done https://github.com/opf/openproject/pull/19449/commits/5a9a0f95e82ca13bfe72923dc8847e3eeb402700

toy commented on spec/models/work_package/pdf_export/common/macro_spec.rb:
```diff
@@ -31,14 +31,60 @@
 require "spec_helper"
 
 RSpec.describe WorkPackage::PDFExport::Common::Macro do
```
Probably better to check result for formatted custom fields too

as-op replied:
done https://github.com/opf/openproject/pull/19449/commits/d75e6a50b6d8a91986635674462661573f52550c

as-op committed:
use consistent naming

resolves https://github.com/opf/openproject/pull/19449#discussion_r2195313118

as-op committed:
better naming

resolves https://github.com/opf/openproject/pull/19449#discussion_r2195349876

as-op committed:
better naming & parameter

resolves https://github.com/opf/openproject/pull/19449#discussion_r2195351784

as-op committed:
remove rescue I18n::MissingTranslationData as it's thrown in dev & test; but disable it for the tests it would be thrown

resolves https://github.com/opf/openproject/pull/19449#discussion_r2195367016

as-op committed:
refactor call

resolves https://github.com/opf/openproject/pull/19449#discussion_r2195515426

as-op committed:
do not use static work package id, but add interpolation in all usages

resolves https://github.com/opf/openproject/pull/19449#discussion_r2195667995

as-op committed:
more compact boolean complexity

resolves https://github.com/opf/openproject/pull/19449#discussion_r2195518040

as-op committed:
do not return an empty string

resolves https://github.com/opf/openproject/pull/19449#discussion_r2195520084

as-op committed:
use the automatically built project_phase_definition

resolves https://github.com/opf/openproject/pull/19449#discussion_r2195529232

as-op committed:
remove not needed save! on work_package and project

resolves https://github.com/opf/openproject/pull/19449#discussion_r2195602347 and https://github.com/opf/openproject/pull/19449#discussion_r2195602900

as-op committed:
use single quotes to avoid escaping double quotes

resolves https://github.com/opf/openproject/pull/19449#discussion_r2195649689

as-op committed:
remove unnecessary new lines in the test

resolves https://github.com/opf/openproject/pull/19449#discussion_r2195651639

as-op committed:
add test for unsupported formatted Custom Field

resolves https://github.com/opf/openproject/pull/19449#discussion_r2195651639

as-op committed:
add tests with another work package referenced by ID

resolves https://github.com/opf/openproject/pull/19449#discussion_r2195656739

as-op committed:
add tests with another project referenced by ID

resolves https://github.com/opf/openproject/pull/19449#discussion_r2195669032

as-op committed:
add tests with restricted project/work_package referenced by ID

as-op committed:
Merge branch 'release/16.2' into bug/65401-project-phase-field-shows-in-pdf-report-even-when-it-doesn-t-show-on-wp-i-e-when-project-phases-are-not-active

as-op committed:
Merge branch 'release/16.2' into bug/65401-project-phase-field-shows-in-pdf-report-even-when-it-doesn-t-show-on-wp-i-e-when-project-phases-are-not-active

as-op committed:
Merge branch 'release/16.2' into bug/65401-project-phase-field-shows-in-pdf-report-even-when-it-doesn-t-show-on-wp-i-e-when-project-phases-are-not-active
